### PR TITLE
Enable --help for finetune CLI

### PR DESCRIPTION
**Pull request by quic-mamta**

### CODE REVIEW

File: QEfficient/finetune/configs/training.py [hunk 1]
```diff
@@ -16,7 +16,8 @@ class TrainConfig:
         model_name (str): Name of the pre-trained model to fine-tune (default: "meta-llama/Llama-3.2-1B").
         tokenizer_name (str): Name of the tokenizer (defaults to model_name if None).
         run_validation (bool): Whether to run validation during training (default: True).
-        batch_size_training (int): Batch size for training (default: 1).
+        train_batch_size (int): Batch size for training (default: 1).
```
Comment (quic-meetkuma) [inline QEfficient/finetune/configs/training.py:19]: good one

File: QEfficient/finetune/configs/training.py [hunk 2]
```diff
@@ -58,7 +59,8 @@ class TrainConfig:
     model_name: str = "meta-llama/Llama-3.2-1B"
     tokenizer_name: str = None  # if not passed as an argument, it uses the value of model_name
     run_validation: bool = True
-    batch_size_training: int = 1
+    train_batch_size: int = 1
```
Comment (quic-meetkuma) [inline QEfficient/finetune/configs/training.py:62]: if we are changing this param, may be an internal announcement is required. SIT team's testing commands might depend on this one.

File: QEfficient/finetune/utils/parser.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    parser.add_argument(
+        "--dataset", required=False, type=str, default="samsum_dataset", help="Dataset name for finetuning"
+    )
+    parser.add_argument(
+        "--task-type",
+        "--task_type",
+        required=False,
+        type=str,
+        default="generation",
+        help="generation/seq_classification",
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/parser.py:120]: Task used for finetuning. Use 'generation' for decoder based models and 'seq_classification' for encoder based models.

File: QEfficient/finetune/utils/parser.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    )
+    parser.add_argument(
+        "--peft-method",
+        "--peft_method",
+        required=False,
+        type=str,
+        default="lora",
+        help="Parameter-efficient fine-tuning method",
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/parser.py:128]: Parameter efficient finetuning technique to be used. Available options: <list_all_available_options>

File: QEfficient/finetune/utils/parser.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        "--output-dir",
+        "--output_dir",
+        required=False,
+        type=str,
+        default="meta-llama-samsum",
+        help="Directory to save outputs",
+    )
+    parser.add_argument(
+        "--num-freeze-layers",
+        "--num_freeze_layers",
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/parser.py:156]: Are we using this argument in code?

File: QEfficient/finetune/utils/parser.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        default=1000,
+        help="Steps between intermediate saves",
+    )
+    parser.add_argument(
+        "--batching-strategy",
+        "--batching_strategy",
+        required=False,
+        type=str,
+        default="packing",
+        help="Batching strategy",
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/parser.py:187]: add some description as well. What packing is and what padding is?

I think this option is not even used in the code. By default we are padding.

File: QEfficient/finetune/utils/parser.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+    parser.add_argument(
+        "--dist-backend",
+        "--dist_backend",
+        required=False,
+        type=str,
+        default="cpu:gloo,qaic:qccl,cuda:gloo",
+        help="Backend for distributed training",
+    )
+    parser.add_argument(
+        "--grad-scaler", "--grad_scaler", required=False, type=bool, default=True, help="Use gradient scaler"
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/parser.py:228]: this should be by default. We are anyways not exposing --use_fp16. That means we need to use grad_scaler. Please correct me if I am wrong.

File: QEfficient/finetune/utils/parser.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+    )
+    parser.add_argument(
+        "--dump-root-dir",
+        "--dump_root_dir",
+        required=False,
+        type=str,
+        default="meta-llama-samsum-mismatches/step_",
+        help="Directory for mismatch dumps",
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/parser.py:236]: Add some more details

File: QEfficient/finetune/utils/train_utils.py [hunk 9]
```diff
@@ -469,14 +469,6 @@ def get_longest_seq_length(data: List[Dict]) -> Tuple[int, int]:
     return longest_seq_length, longest_seq_ix
 
 
-def get_parameter_dtypes(model):
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/train_utils.py:472]: Is this removed because it is not used anywhere?

File: QEfficient/finetune/utils/dataset_utils.py [hunk 10]
```diff
@@ -29,20 +28,3 @@ def get_custom_data_collator(dataset_processer, dataset_config) -> torch.utils.d
         return None
 
     return DATALOADER_COLLATE_FUNC[dataset_config.dataset](dataset_processer, dataset_config)
-
-
-def get_dataloader(tokenizer, dataset_config, train_config, split: str = "train"):
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/dataset_utils.py:34]: Why this function was removed?

Is this removed because it is not used anywhere?

**Review by quic-meetkuma — CHANGES_REQUESTED** [review]
Overall good cleanup. Please check comments.

File: QEfficient/finetune/utils/train_utils.py (hunk 9 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/train_utils.py:472]: yes it is redundant code.

File: QEfficient/finetune/utils/parser.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        required=False,
+        type=float,
+        default=1e-4,
+        help="Loss threshold for convergence, if loss value is <= convergence_loss for #convergence_counter consecutive steps, fine tuning stops",
+    )
+    parser.add_argument(
+        "--use-profiler", "--use_profiler", required=False, type=bool, default=False, help="Enable profiling"
+    )
+    parser.add_argument(
+        "--enable-ddp", "--enable_ddp", required=False, type=int, default=1000, help="Enable distributed data parallel"
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/parser.py:217]: Why we need this flag? Can we not determine from environment flags?

ususally user is already passing CLI as "torchrun --nproc-per-node <num_devices> -m QEfficient.cloud.finetune"

when running in ddp model os.getenv('WORLD_SIZE') should give number of devices. In case of non-ddp this variable is None.

File: QEfficient/finetune/utils/parser.py (hunk 3 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/parser.py:120]: updated in latest.

File: QEfficient/finetune/utils/parser.py (hunk 5 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/parser.py:156]: it's an unused variable.

File: QEfficient/finetune/configs/training.py [hunk 12]
```diff
@@ -29,17 +30,17 @@ class TrainConfig:
         weight_decay (float): Weight decay for optimizer (default: 0.0).
         gamma (float): Learning rate decay factor (default: 0.85).
         seed (int): Random seed for reproducibility (default: 42).
+        use_autocast (bool): Use autocast for mixed precision (default: True).
+        dataset (str): Dataset name for training (default: "samsum_dataset").
         use_fp16 (bool): Use mixed precision training (default: True).
```
[the remark this answers is not in the document]
Comment (quic-mamta) [reply QEfficient/finetune/configs/training.py:35]: done.

File: QEfficient/finetune/utils/parser.py (hunk 4 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/parser.py:128]: for this there are no other choices, currently we only support lora.

File: QEfficient/cloud/finetune.py [hunk 13]
```diff
@@ -354,4 +354,6 @@ def main(peft_config_file: str = None, **kwargs) -> None:
 
 
 if __name__ == "__main__":
-    fire.Fire(main)
+    parser = get_finetune_parser()
+    args = parser.parse_args()
+    main(**args.__dict__)
```
[the remark this answers is not in the document]
Comment (quic-mamta) [reply QEfficient/cloud/finetune.py:359]: updated in latest

File: QEfficient/finetune/utils/parser.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+        "--use-autocast",
+        "--use_autocast",
+        required=False,
+        type=bool,
+        default=True,
+        help="Use autocast for mixed precision",
+    )
+
+    parser.add_argument(
+        "--dataset", required=False, type=str, default="samsum_dataset", help="Dataset name for finetuning"
```
[the remark this answers is not in the document]
Comment (quic-mamta) [reply QEfficient/finetune/utils/parser.py:112]: updated in latest

File: QEfficient/finetune/utils/parser.py (hunk 7 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/parser.py:228]: updated

File: QEfficient/finetune/utils/parser.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (quic-mamta) [reply QEfficient/finetune/utils/parser.py]: this is a big change, will take it in next PR.

File: QEfficient/finetune/eval.py [hunk 15]
```diff
@@ -42,18 +36,20 @@ def main(**kwargs):
...
 
     # Set the seeds for reproducibility
     torch.manual_seed(train_config.seed)
     random.seed(train_config.seed)
     np.random.seed(train_config.seed)
 
     # Load the pre-trained model and setup its configuration
-    # config = AutoConfig.from_pretrained(train_config.model_name)
-    save_dir = "meta-llama-samsum/trained_weights/step_14000"
+    save_dir = os.path.join(train_config.output_dir, "complete_epoch_1")
```
Comment (quic-meetkuma) [inline QEfficient/finetune/eval.py:48]: complete_epoch_1 should not be hard coded. Instead we should identify the latest checkpoint and load that.

File: QEfficient/finetune/utils/parser.py [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+        choices=DATASET_PREPROC.keys(),
+        help="Dataset name to be used for finetuning (default: %(default)s)",
+    )
+    parser.add_argument(
+        "--task_type",
+        "--task-type",
+        required=False,
+        type=str,
+        default="generation",
+        choices=["generation", "seq_classification"],
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/parser.py:121]: dont hard code here. Make a variable and use it everywhere in the code.

File: QEfficient/finetune/utils/parser.py [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+        help="Task used for finetuning. Use 'generation' for decoder based models and 'seq_classification' for encoder based models.",
+    )
+    parser.add_argument(
+        "--peft_method",
+        "--peft-method",
+        required=False,
+        type=str,
+        default="lora",
+        choices=["lora"],
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/parser.py:130]: same comment as --task_type

File: QEfficient/finetune/utils/parser.py [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+        type=str,
+        default="",
+        help="Path to load PEFT checkpoint and resume the fine-tuning on that checkpoint",
+    )
+    parser.add_argument(
+        "--output_dir",
+        "--output-dir",
+        required=False,
+        type=str,
+        default="meta-llama-samsum",
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/parser.py:146]: this should change. can we make it generic e.g. training_results

File: QEfficient/finetune/utils/train_utils.py [hunk 19]
```diff
@@ -123,7 +121,7 @@ def train(
                 )
                 break
 
-        if train_config.use_peft and train_config.from_peft_checkpoint:
+        if train_config.peft_method == "lora" and train_config.from_peft_checkpoint:
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/train_utils.py:124]: instead of matching the string, we can make 'lora' as enum and compare against it. May be we can take such changes in next PR. The same should be applied to task type.

File: QEfficient/finetune/utils/train_utils.py [hunk 20]
```diff
@@ -173,9 +171,7 @@ def train(
                 break
             batch = {k: v.to(device) for k, v in batch.items()}  # move the batch elements to qaic device
 
-            with (
-                torch.autocast(device_type=device, dtype=torch.float16) if train_config.use_autocast else nullcontext()
-            ):
+            with torch.autocast(device_type=device, dtype=torch.float16):
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/train_utils.py:174]: ideally it should be torch.device(device).type so that "cuda:0" can be treated as "cuda" and "cuda" is also treated as "cuda".

File: QEfficient/finetune/utils/train_utils.py [hunk 21]
```diff
@@ -427,9 +417,7 @@ def evaluation_helper(model, train_config, eval_dataloader, device):
         # Ensure no gradients are computed for this scope to save memory
         with torch.no_grad():
             # Forward pass and compute loss
-            with (
-                torch.autocast(device_type=device, dtype=torch.float16) if train_config.use_autocast else nullcontext()
-            ):
+            with torch.autocast(device_type=device, dtype=torch.float16):
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/train_utils.py:420]: same comment as above

File: QEfficient/cloud/finetune.py [hunk 22]
```diff
@@ -68,7 +64,8 @@ def setup_distributed_training(train_config: TrainConfig) -> None:
     assert torch_device.type != "cpu", "Host doesn't support single-node DDP"
     assert torch_device.index is None, f"DDP requires only device type, got: {torch_device}"
 
-    dist.init_process_group(backend=train_config.dist_backend)
+    dist_backend = {"cpu": "gloo", "qaic": "qccl", "cuda": "gloo"}
+    dist.init_process_group(backend=dist_backend[torch_device.type])
```
Comment (quic-meetkuma) [inline QEfficient/cloud/finetune.py:68]: first check whether the user specified backend is present in dist_backend or not?

**Review by quic-meetkuma — CHANGES_REQUESTED** [review]
Overall good cleanup.

File: QEfficient/finetune/utils/train_utils.py (hunk 20 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/train_utils.py:174]: updated.

File: QEfficient/finetune/utils/train_utils.py (hunk 21 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/train_utils.py:420]: done.

File: QEfficient/finetune/utils/parser.py (hunk 18 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/parser.py:146]: done.

File: QEfficient/finetune/utils/parser.py (hunk 16 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/parser.py:121]: done.

File: QEfficient/finetune/utils/parser.py (hunk 17 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/parser.py:130]: done.

File: QEfficient/finetune/utils/train_utils.py (hunk 19 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/train_utils.py]: updated.

File: QEfficient/finetune/utils/parser.py [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+
+import argparse
+
+from QEfficient.finetune.dataset.dataset_config import DATASET_PREPROC
+from QEfficient.finetune.utils.helper import BATCHING_STRATEGY, DEVICE, PEFT_METHOD, TASK_TYPE
+
+
+def get_finetune_parser():
+    parser = argparse.ArgumentParser(
+        description="Finetune command, the model will be downloaded from Huggingface and finetuned on Cloud AI 100 and weights are saved."
```
Comment (quic-swatia) [inline QEfficient/finetune/utils/parser.py:16]: Please make 'and then checkpoints will be saved in the output_dir'.

File: QEfficient/finetune/configs/training.py [hunk 24]
```diff
@@ -98,10 +88,5 @@ class TrainConfig:
     use_profiler: bool = False  # Enable pytorch profiler, can not be used with flop counter at the same time.
     # profiler_dir: str = "PATH/to/save/profiler/results" # will be used if using profiler
 
-    # dist-related
-    enable_ddp: bool = False
-    dist_backend: str = "cpu:gloo,qaic:qccl,cuda:gloo"
-
-    grad_scaler: bool = True
```
Comment (quic-swatia) [inline QEfficient/finetune/configs/training.py:105]: Please restore 'grad_scaler' and 'use_autocast' flags as discussed.

File: QEfficient/cloud/finetune.py [hunk 25]
```diff
@@ -226,58 +224,13 @@ def setup_dataloaders(
...
-        if custom_data_collator:
-            val_dl_kwargs["collate_fn"] = custom_data_collator
-
-        eval_dataloader = torch.utils.data.DataLoader(
-            dataset_val,
-            num_workers=train_config.num_workers_dataloader,
-            pin_memory=True,
-            **val_dl_kwargs,
-        )
+        eval_dataloader = get_dataloader(tokenizer, dataset_config, train_config, split="test")
```
[the remark this answers is not in the document]
Comment (quic-mamta) [reply QEfficient/cloud/finetune.py:233]: done

File: QEfficient/finetune/configs/training.py (hunk 24 above)
Comment (quic-mamta) [reply QEfficient/finetune/configs/training.py:105]: done

**Review by quic-swatia — APPROVED** [review]

**Review by vbaddi — APPROVED** [review]
